transport-native-kqueue/src/main/c/netty_kqueue_native.c

```diff
@@ -0,0 +1,307 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+#include <errno.h>
+#include <netdb.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <dlfcn.h>
+#include <stddef.h>
+#include <time.h>
+#include <sys/event.h>
+#include <sys/socket.h>
+#include <sys/time.h>
+#include "netty_unix_filedescriptor.h"
+#include "netty_unix_socket.h"
+#include "netty_unix_errors.h"
+#include "netty_unix_util.h"
+#include "netty_unix_limits.h"
+#include "netty_kqueue_bsdsocket.h"
+#include "netty_kqueue_eventarray.h"
+
+clockid_t waitClockId = 0; // initialized by netty_unix_util_initialize_wait_clock
+
+static jint netty_kqueue_native_kqueueCreate(JNIEnv* env, jclass clazz) {
+    jint kq = kqueue();
+    if (kq < 0) {
+      netty_unix_errors_throwChannelExceptionErrorNo(env, "kqueue() failed: ", errno);
+    }
+    return kq;
+}
+
+static jint netty_kqueue_native_keventChangeSingleUserEvent(jint kqueueFd, struct kevent* userEvent) {
+    int result, err;
+    result = kevent(kqueueFd, userEvent, 1, NULL, 0, NULL);
+    if (result < 0) {
+        // https://www.freebsd.org/cgi/man.cgi?query=kqueue&sektion=2
+        // When kevent() call fails with EINTR error, all changes in the changelist have been applied.
+        return (err = errno) == EINTR ? 0 : -err;
+    }
+
+    return result;
+}
+
+static jint netty_kqueue_native_keventTriggerUserEvent(JNIEnv* env, jclass clazz, jint kqueueFd, jint ident) {
+    struct kevent userEvent;
+    EV_SET(&userEvent, ident, EVFILT_USER, 0, NOTE_TRIGGER | NOTE_FFNOP, 0, NULL);
+    return netty_kqueue_native_keventChangeSingleUserEvent(kqueueFd, &userEvent);
+}
+
+static jint netty_kqueue_native_keventAddUserEvent(JNIEnv* env, jclass clazz, jint kqueueFd, jint ident) {
+    struct kevent userEvent;
+    EV_SET(&userEvent, ident, EVFILT_USER, EV_ADD | EV_ENABLE | EV_CLEAR, NOTE_FFNOP, 0, NULL);
+    return netty_kqueue_native_keventChangeSingleUserEvent(kqueueFd, &userEvent);
+}
+
+static jint netty_kqueue_native_keventWait(JNIEnv* env, jclass clazz, jint kqueueFd, jlong changeListAddress, jint changeListLength,
+                                           jlong eventListAddress, jint eventListLength, jint tvSec, jint tvNsec) {
+    struct kevent* changeList = (struct kevent*) changeListAddress;
+    struct kevent* eventList = (struct kevent*) eventListAddress;
+    struct timespec beforeTs, nowTs, timeoutTs;
+    int result, err;
+
+    timeoutTs.tv_sec = tvSec;
+    timeoutTs.tv_nsec = tvNsec;
+
+    // Negatives = wait indefinitely, Zeros = poll (aka return immediately).
+    if ((tvSec == 0 && tvNsec == 0) || tvSec < 0 || tvNsec < 0) {
```

Comment:
@normanmaurer - note that I added support to wait indefinitely (negative time values), but this is not currently used. If we decide that all tasks should wakeup the event loop (e.g. writes from outside the event loop) then we may be able to utilize this. It makes the code a bit more defensive and covers all potential cases, but I can remove it if you think its overkill (we can always add it later). We currently don't check for negative values (or zero) in EPOLL.

In the case of negative (wait forever) or zero (poll now) we can avoid getting the system time which may be a relatively costly operation depending on the clock. We may want to consider at least the zero check for EPOLL too.